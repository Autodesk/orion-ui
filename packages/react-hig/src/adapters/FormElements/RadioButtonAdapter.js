/*
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import * as HIG from 'hig.web';

import HIGElement from '../../elements/HIGElement';
import * as PropTypes from 'prop-types';
import createComponent from '../createComponent';

export class RadioButtonAdapter extends HIGElement {
  constructor(initialProps) {
    super(HIG.RadioButton, initialProps);
  }

  componentDidMount() {
    if (this.props.required) {
      this.hig.required();
    }
    if (this.props.checked) {
      this.hig.check();
    }
    if (this.props.disabled) {
      this.hig.disable();
    }
  }

  commitUpdate(updatePayload, oldProps, newProps) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'label':
          this.hig.setLabel(propValue);
          break;
        case 'name':
          this.hig.setName(propValue);
          break;
        case 'value':
          this.hig.setValue(propValue);
          break;
        case 'required':
          propValue ? this.hig.required() : this.hig.noLongerRequired();
          break;
        case 'checked':
          propValue ? this.hig.check() : this.hig.uncheck();
          break;
        case 'disabled':
          propValue ? this.hig.disable() : this.hig.enable();
          break;
        case 'onChange': {
          const dispose = this._disposeFunctions.get('onChangeDispose');
          if (dispose) {
            dispose();
          }
          this._disposeFunctions.set(
            'onChangeDispose',
            this.hig.onChange(propValue)
          );
          break;
        }
        case 'onFocus': {
          const dispose = this._disposeFunctions.get('onFocusDispose');
          if (dispose) {
            dispose();
          }
          this._disposeFunctions.set(
            'onFocusDispose',
            this.hig.onFocus(propValue)
          );
          break;
        }
        case 'onHover': {
          const dispose = this._disposeFunctions.get('onHoverDispose');
          if (dispose) {
            dispose();
          }
          this._disposeFunctions.set(
            'onHoverDispose',
            this.hig.onHover(propValue)
          );
          break;
        }
        default:
          console.warn('Unknown key, not handled: ', propKey);
      }
    }
  }
}

const RadioButtonAdapterComponent = createComponent(RadioButtonAdapter);

RadioButtonAdapterComponent.propTypes = {
  disabled: PropTypes.bool,
  checked: PropTypes.bool,
  required: PropTypes.bool,
  onHover: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  name: PropTypes.string,
  value: PropTypes.string,
  label: PropTypes.string
};

RadioButtonAdapterComponent.__docgenInfo = {
  props: {
    name: {
      description: 'sets the name attribute of the radiobutton input'
    },
    value: {
      description: 'sets the value attribute of the radiobutton input'
    },
    checked: {
      description: 'boolean - sets whether the radiobutton is checked'
    },
    disabled: {
      description: 'boolean - sets whether the radiobutton is disabled'
    },
    required: {
      description: 'boolean - sets the whether the radiobutton is required'
    },
    label: {
      description: 'sets the label text for the radiobutton'
    },
    onHover: {
      description: 'triggers when you hover over the button'
    },
    onChange: {
      description: 'triggers when you check/uncheck the radiobutton'
    },
    onFocus: {
      description: 'triggers the radiobutton component receives focus'
    }
  }
};

export default RadioButtonAdapterComponent;
