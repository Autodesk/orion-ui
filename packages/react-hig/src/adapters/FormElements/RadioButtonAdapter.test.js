/*
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */
import { mount } from 'enzyme';
import * as HIG from 'hig.web';
import React from 'react';

import { default as RadioButton } from './RadioButtonAdapter';

const inputId = '1234';

describe('<RadioButton>', () => {
  function createHigComponent(defaults = {}) {
    const higContainer = document.createElement('div');

    // use spread here to clone defaults since HIG.Checkbox mutates this object
    const higRadioButton = new HIG.RadioButton({ ...defaults });

    higRadioButton.mount(higContainer);

    // to adjust for the randomly generated id
    const label = higContainer.querySelector('label');
    const input = higContainer.querySelector('input');
    label.setAttribute('for', inputId);
    input.setAttribute('id', inputId);

    return { higComponent: higRadioButton, higContainer };
  }

  function createOrionComponent(defaults) {
    const orionContainer = document.createElement('div');
    const orionWrapper = mount(<RadioButton {...defaults} />, {
      attachTo: orionContainer
    });

    const label = orionContainer.querySelector('label');
    const input = orionContainer.querySelector('input');

    // to adjust for the randomly generated id
    label.setAttribute('for', inputId);
    input.setAttribute('id', inputId);
    return { orionWrapper, orionContainer };
  }

  it('renders', () => {
    const defaults = {};
    const { higComponent, higContainer } = createHigComponent(defaults);
    const { orionContainer, orionWrapper } = createOrionComponent(defaults);

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders with initial props', () => {
    const defaults = {
      label: 'Label',
      name: 'some-ame',
      value: 'some-value'
    };
    const { higComponent, higContainer } = createHigComponent(defaults);
    const { orionContainer, orionWrapper } = createOrionComponent(defaults);

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders with updated props', () => {
    const defaults = {};
    const nextProps = {
      label: 'Label',
      name: 'some-name',
      value: 'some-value',
      checked: true,
      required: 'You must make a selection',
      disabled: true
    };
    const { higComponent, higContainer } = createHigComponent(defaults);
    const { orionContainer, orionWrapper } = createOrionComponent(defaults);

    orionWrapper.setProps(nextProps);

    higComponent.setLabel(nextProps.label);
    higComponent.setName(nextProps.name);
    higComponent.setValue(nextProps.value);
    higComponent.check();
    higComponent.required(nextProps.required);
    higComponent.disable();

    expect(orionContainer.firstElementChild.outerHTML).toMatchSnapshot();

    expect(orionContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  ['onChange', 'onFocus', 'onHover'].forEach(eventName => {
    it(`sets event listeners for ${eventName} initially`, () => {
      const { orionContainer, orionWrapper } = createOrionComponent({
        [eventName]: () => {}
      });
      const instance = orionWrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(eventName);
      expect(disposeFunction).toBeDefined();
    });

    it(`sets event listeners for ${eventName} when updated`, () => {
      const { orionContainer, orionWrapper } = createOrionComponent({
        [eventName]: () => {}
      });
      orionWrapper.setProps({ [eventName]: () => {} });
      const instance = orionWrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(
        `${eventName}Dispose`
      );
      expect(disposeFunction).toBeDefined();
    });
  });
});
