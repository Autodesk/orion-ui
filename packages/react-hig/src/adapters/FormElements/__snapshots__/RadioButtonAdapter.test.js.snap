// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButton> does not show a label if not specified 1`] = `
"<div class=\\"hig__input-button hig__input-button--radio \\">
  <!--RADIOBUTTON_LABEL-->
  <label for=\\"1234\\" class=\\"hig__input-button__label hig--hidden\\"></label>
  <input type=\\"radio\\" class=\\"hig__input-button__input\\" name=\\"agree_toc\\" id=\\"1234\\" value=\\"agree\\">
  <span class=\\"hig__input-button__input-wrapper\\"></span>
</div>"
`;

exports[`<RadioButton> renders the standard  RadioButton 1`] = `
"<div class=\\"hig__input-button hig__input-button--radio \\">
  <!--RADIOBUTTON_LABEL-->
  <label for=\\"1234\\" class=\\"hig__input-button__label\\">I agree</label>
  <input type=\\"radio\\" class=\\"hig__input-button__input\\" name=\\"agree_toc\\" id=\\"1234\\" value=\\"agree\\">
  <span class=\\"hig__input-button__input-wrapper\\"></span>
</div>"
`;
